baserun/instrumentation/openai.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python
+import json
+import logging
+from typing import Collection, Any
+
+import openai
+from openai.openai_object import OpenAIObject
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.utils import (
+    unwrap,
+)
+from opentelemetry.sdk.trace import Span
+from opentelemetry.trace import get_tracer
+from wrapt import wrap_function_wrapper
+
+from baserun.instrumentation.span_attributes import SpanAttributes, OPENAI_VENDOR_NAME
+from baserun.instrumentation.wrappers import instrumented_wrapper
+
+logger = logging.getLogger(__name__)
+
+_instruments = ("openai >= 0.27.0",)
+__version__ = "0.1.0"
+
+WRAPPED_METHODS = [
+    {
+        "object": "ChatCompletion",
+        "method": "create",
+        "span_name": "openai.chat",
+    },
+    {
+        "object": "Completion",
+        "method": "create",
+        "span_name": "openai.completion",
+    },
+]
+
+
+class OpenAIInstrumentor(BaseInstrumentor):
+    """An instrumentor for OpenAI's client library."""
+
+    def instrumentation_dependencies(self) -> Collection[str]:
+        return _instruments
+
+    def set_request_attributes(self, span: Span, kwargs: dict[str, Any]):
+        span.set_attribute(SpanAttributes.LLM_VENDOR, OPENAI_VENDOR_NAME)
+        span.set_attribute(SpanAttributes.OPENAI_API_BASE, openai.api_base)
+        span.set_attribute(SpanAttributes.OPENAI_API_TYPE, openai.api_type)
+
+        span.set_attribute(SpanAttributes.LLM_TEMPERATURE, kwargs.get("temperature", 1))
+        span.set_attribute(SpanAttributes.LLM_TOP_P, kwargs.get("top_p", 1))
+        span.set_attribute(
+            SpanAttributes.LLM_FREQUENCY_PENALTY,
+            kwargs.get("frequency_penalty", 0),
+        )
+        span.set_attribute(
+            SpanAttributes.LLM_PRESENCE_PENALTY,
+            kwargs.get("presence_penalty", 0),
+        )
+
+        if functions := kwargs.get("functions"):
+            span.set_attribute(SpanAttributes.LLM_FUNCTIONS, json.dumps(functions))
+
+        if function_call := kwargs.get("function_call"):
+            span.set_attribute(
+                SpanAttributes.LLM_FUNCTION_CALL,
+                json.dumps(function_call),
+            )
+
+        span.set_attribute(SpanAttributes.LLM_N, kwargs.get("n", 1))
+        span.set_attribute(SpanAttributes.LLM_STREAM, kwargs.get("stream", False))
+
+        if stop := kwargs.get("stop"):
+            span.set_attribute(SpanAttributes.LLM_STOP, str(stop))
```

Comment:
I think stop is string | string[], so not sure str() will work for the array case?